### PR TITLE
kubelet: Pull the pod infra container creation logic out.

### PR DESCRIPTION
**Pull request by yifan-gu**
This PR demuxes the infra container logic from the check-hash-and-health-loop of  `computePodContainerChanges`, this enables us to reuse the loop for other runtime.

I think we also want to throw the infra container check logic into a separate function, (let's call it `P()`) which returns false if we don't need to continue to compute the hash/health. So for other runtime, we can just skip `P()` (or make each runtime implement their own version of `P()`)

I haven't come up with a good name for `P()` so I decide to do it in next PR so the review of this one can be smoother.

@vmarmol @dchen1107 @yujuhong

### CODE REVIEW

**Comment by yifan-gu** [comment]
My final goal for `computePodContainerChanges` is to return just `containersToKeep` and `containersToStart`, and let the docker runtime create pod infra container itself. 
This requires moving the `createPodInfraContainer` into dockertools as well.

Feedback appreciate! Thanks!

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -966,6 +966,10 @@ func (kl *Kubelet) makePodDataDirs(pod *api.Pod) error {
 	return nil
 }
 
+// shouldContainerBeRestarted checks for the non-running containers and determines whether they should be
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: nit: checks the non-running containers

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -966,6 +966,10 @@ func (kl *Kubelet) makePodDataDirs(pod *api.Pod) error {
...
+// restart based on the exit state and restart policy. Currently if the spec changes after the container
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: if "for a restart never" container

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1103,6 +1107,42 @@ func (kl *Kubelet) computePodContainerChanges(pod *api.Pod, runningPod kubeconta
 		return podContainerChangesSpec{}, err
 	}
 
+	// If we need to restart pod infra container, then we don't need to probe and check each
+	// container below because all containers will need to be restarted, except for those have
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: except for those that specify a never restart policy

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -1103,6 +1107,42 @@ func (kl *Kubelet) computePodContainerChanges(pod *api.Pod, runningPod kubeconta
...
+	// run successfully.
+	if createPodInfraContainer {
+		// If restart policy is never, then we don't need to keep any containers,
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: We already documented this above

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -1103,6 +1107,42 @@ func (kl *Kubelet) computePodContainerChanges(pod *api.Pod, runningPod kubeconta
...
+		// including the pod infra container.
+		// Note that containersToKeep and containersToStart are empty at this point.
+		if pod.Spec.RestartPolicy == api.RestartPolicyNever {
+			createPodInfraContainer = false
+		} else {
+			for index, container := range pod.Spec.Containers {
+				// Here are the cases where the container should be restarted:
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: Do we need this comment?

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (yifan-gu) [inline pkg/kubelet/kubelet.go]: ack

**Comment by yifan-gu** [comment]
Put off this PR for that we are moving the whole `computePodContainerChanges` to docker container manager.

**Pull request closed without merging**